### PR TITLE
topic/289-typescript

### PR DESCRIPTION
**Pull request by lfrost**
This includes the basic wiring for TypeScript and the conversion of one JavaScript file to TypeScript (as a first step for issue 289).

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -21,6 +21,21 @@ module.exports = {
...
+      files: ['./addon/**/*.ts'],
+      parser: '@typescript-eslint/parser',
+      plugins: ['@typescript-eslint'],
+      extends: ['plugin:@typescript-eslint/recommended'],
+      rules: {
+        semi: ['error', 'never'],
+        'prettier/prettier': [
+          'error',
+          {
+            semi: false,
```
Comment (jherdman) [inline .eslintrc.js:34]: Let's not do this :)

File: package.json [hunk 2]
```diff
@@ -36,25 +37,53 @@
...
+    "@types/ember__utils": "^3.16.2",
+    "@types/htmlbars-inline-precompile": "^1.0.1",
+    "@types/qunit": "^2.11.2",
+    "@types/rsvp": "^4.0.4",
+    "@typescript-eslint/eslint-plugin": "^4.32.0",
+    "@typescript-eslint/parser": "^4.32.0",
     "babel-eslint": "^10.1.0",
     "broccoli-asset-rev": "^3.0.0",
-    "ember-cli": "~3.28.0",
+    "ember-cli": "~3.27.0",
```
Comment (jherdman) [inline package.json:80]: This seems like a mistake.

**Review by jherdman — CHANGES_REQUESTED** [review]
Thank you for doing this. Just a few comments above.

File: .eslintrc.js (hunk 1 above)
Comment (lfrost) [reply .eslintrc.js]: Whichever way you prefer.  I pushed a change for this.

File: package.json (hunk 2 above)
Comment (lfrost) [reply package.json:80]: No, it is a workaround for a [bug in ember-cli 3.28.0](https://github.com/ember-cli/ember-cli/issues/9606).  Now that ember-cli-typescript is installed, we could upgrade back to 3.28.0, but I would wait for 3.28.1 to be released.

Comment (jherdman) [reply package.json:80]: Just this change left.

**Review by jherdman — CHANGES_REQUESTED** [review]
One last note.

File: .gitignore [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 # compiled output
 /dist/
+/doc/
```
[the remark this answers is not in the document]
Comment (lfrost) [reply .gitignore:5]: Yes.  I did not notice the project had a .npmignore.  I have pushed this change.

**Review by jherdman — APPROVED** [review]

**Comment by jherdman** [comment]
Many thanks for your contribution! 😍

**Comment by jherdman** [comment]
Resolves #289

**Pull request merged**